src/mongo_logger.ts

```diff
@@ -0,0 +1,197 @@
+ */
+function parseMaxDocumentLength(s?: string): number {
+  try {
+    const maxDocumentLength = getUint('MONGODB_LOG_MAX_DOCUMENT_LENGTH', s);
+    return maxDocumentLength;
+  } catch {
+    return 1000;
+  }
+}
+
+/**
+ * Parses a string for logDestination. Strings containing 'stderr' or 'stdout' are treated as
+ * case-insensitive.
+ *
+ * @param logDestination - the destination to write logs to
+ * @returns the original string if a file path was passed. Otherwise one of ['stderr', 'stdout'].
+ */
+function parseLogDestination(logDestination?: string): string {
+  if (typeof logDestination === 'string' && logDestination !== '') {
+    const lowerLogDestination = logDestination.toLowerCase();
+    return lowerLogDestination === 'stdout' || lowerLogDestination === 'stderr'
+      ? lowerLogDestination
+      : logDestination;
+  }
+
+  return 'stderr';
+}
+
+/** @internal */
+export const MongoLoggableComponent = Object.freeze({
+  COMMAND: 'command',
+  TOPOLOGY: 'topology',
+  SERVER_SELECTION: 'serverSelection',
+  CONNECTION: 'connection'
+} as const);
+
+/** @internal */
+export type MongoLoggableComponent =
+  typeof MongoLoggableComponent[keyof typeof MongoLoggableComponent];
+
+/** @internal */
+export interface MongoLoggerEnvOptions {
+  /** Severity level for command component */
+  MONGODB_LOG_COMMAND?: string;
+  /** Severity level for topology component */
+  MONGODB_LOG_TOPOLOGY?: string;
+  /** Severity level for server selection component */
+  MONGODB_LOG_SERVER_SELECTION?: string;
+  /** Severity level for CMAP */
+  MONGODB_LOG_CONNECTION?: string;
+  /** Default severity level to be if any of the above are unset */
+  MONGODB_LOG_ALL?: string;
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  MONGODB_LOG_MAX_DOCUMENT_LENGTH?: string;
+  /** Destination for log messages. Must be 'stderr', 'stdout', or a file path. Defaults to 'stderr'. */
+  MONGODB_LOG_PATH?: string;
+}
+
+/** @internal */
+export interface MongoLoggerMongoClientOptions {
+  /** Destination for log messages. Must be 'stderr' or 'stdout'. Defaults to 'stderr'. */
+  mongodbLogPath?: string;
+}
+
+/** @internal */
+export interface MongoLoggerOptions {
+  severitySettings: {
+    /** Severity level for command component */
+    command: SeverityLevel;
+    /** Severity level for topology component */
+    topology: SeverityLevel;
+    /** Severity level for server selection component */
+    serverSelection: SeverityLevel;
+    /** Severity level for connection component */
+    connection: SeverityLevel;
+    /** Default severity level to be used if any of the above are unset */
+    default: SeverityLevel;
+  };
+
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  maxDocumentLength: number;
+  /** Destination for log messages. Must be 'stderr' or 'stdout'. Defaults to 'stderr'. */
+  logDestination: string;
```

Comment:
we need to be able to pipe to a stream - is there a separate ticket to add it? it needs to either be added here, or we need to reference the follow up ticket indicating that the implementation is incomplete; we will not be able to test anything from the spec if we cannot take a custom stream destination